External RPC request whitelist

sebastiandelgado opened the pull request:
Addresses #912

    Create a whitelist that prevents external RPC requests (from dapps) from using methods like shh.setMinPoW. The approach here was to pass down an 'external' flag into the context.Context of call raw, since its a request-scoped variable.

Done
 - backend.rpc:CallRPC calls CallRaw using an external flag
 - All other CallRaw calls are updated to use external=false. Don't think that any of them are external, but correct me if I'm wrong.
 - external flag gets passed into the context.Context of CallRaw
 - When calling each method, the whitelist is applied to external requests, returning an error

sebastiandelgado commented on t/e2e/whisper/whisper_mailbox_test.go:
```diff
@@ -379,10 +379,10 @@ func (s *WhisperMailboxSuite) createPrivateChatMessageFilter(rpcCli *rpc.Client,
 	resp := rpcCli.CallRaw(`{
 			"jsonrpc": "2.0",
 			"method": "shh_newMessageFilter", "params": [
-				{"privateKeyID": "` + privateKeyID + `", "topics": [ "` + topic + `"], "allowP2P":true}
```
The linter in `make ci` insisted that this needed to be gofmt-compliant

mandrigin commented on geth/jail/jail.go:
```diff
@@ -268,7 +268,8 @@ func (j *Jail) sendRPCCall(request string) (interface{}, error) {
 	}
 
-	rawResponse := client.CallRaw(request)
+	isExternal := false
```
I think jail calls are external, actually.

mandrigin commented on geth/rpc/call_raw.go:
```diff
@@ -115,6 +125,12 @@ func (c *Client) callSingleMethod(ctx context.Context, msg json.RawMessage) stri
 
+	// If the req is external, check the whitelist
+	if ctx.Value(externalRPCContextKey) == true && !methodInWhitelist(method) {
+		err = fmt.Errorf("RPC method '%s' not whitelisted for external use", method)
```
I think we should use the same error message as with the method not found. We don't want to give an adversary any info that these methods are event exitsing.

mandrigin reviewed (approved):
LGTM, but I'd keep the mailbox test reformating in a separate commit.

adambabik commented on geth/rpc/call_raw.go:
```diff
@@ -56,6 +69,8 @@ type jsonError struct {
 	Data    interface{} `json:"data,omitempty"`
 }
 
+type contextKey string
```
Can we move it to the top of this file before `const` in which it is used?

sebastiandelgado replied:
Will update this before landing

adambabik commented on geth/rpc/call_raw.go:
```diff
@@ -20,10 +22,21 @@ const (
 // thus, we will use zero ID as a workaround of this limitation
 var defaultMsgID = json.RawMessage(`0`)
 
+var externalRPCMethodWhitelist = []string{
```
@mandrigin are these really all JSON-RPC methods we should allow? INFURA, for example, supports many more `eth_*`.

sebastiandelgado replied:
Those are the rpc methods that were referenced by #912 (personal_sign and personal_ecRecover) plus the rpc methods that needed to be added to pass the mailbox tests (web3_clientVersion and eth_syncing). Where can I look or who can I ask for a more complete listing of the methods that need to be added here? Once I build the app with these changes, are there any good example apps I can run to check if I'm breaking anything with this limited whitelist?

adambabik replied:
Supported methods by INFURA may be a good initial list. Please take a look at [`route.go`](https://github.com/status-im/status-go/blob/develop/geth/rpc/route.go#L35-L88).

> Once I build the app with these changes, are there any good example apps I can run to check if I'm breaking anything with this limited whitelist?

@mandrigin or people in `#core-dapps` channel should be able to help with this.

mandrigin replied:
@adambabik everything else is exported anyway, full namespaces, as normal. So far we really need `personal_sign` and `personal_ecRecover`, but we should not expose the whole `personal_*` namespace.
That was done implicitly, and I'm glad that this PR makes it explicit! :)

adambabik reviewed (commented):
I would recommend building the app with these changes and verifying that it works properly with DApps.

The code looks great! 👏

Commit:
Add isExternal param to CallRaw for rpc whitelist

Add the isExternal param in order to implement the external RPC call whitelist detailed in #912

mandrigin reviewed (changes requested):
Good job!
I think we should still keep some namespaces fully whitelisted like they were before. Essentially the whitelist should only have `personal` namespace method, as far as I concerned.

Otherwise, some of the methods like `eth_newBlockFilter`, and others won't be available anymore.

So we need to either a way to whitelist the whole namespaces here (like "eth_*") or figure out a slightly different mechanics.

sebastiandelgado commented:
@mandrigin Yeah that should be straightforward. We can whitelist whole namespaces (all of them except 'personal') and then whitelist 'personal_ecRecover' and 'personal_sign' individually. Does that sound good? Here's the list of whitespaced namespaces I will use:

- eth
- db
- shh
- web3
- net

I think thats all of them. Am I missing any of them?

mandrigin commented:
@sebastiandelgado sounds good to me!

mandrigin reviewed (approved):
LGTM!

Commit:
External RPC method whitelist
Addresses #912

Create a whitelist that prevents external RPC requests (from dapps) from using methods like shh.setMinPoW. The approach here was to pass down an 'external' flag into the context.Context of call raw, since its a request-scoped variable

sebastiandelgado commented:
Replaced by #967
